src/hotspot/share/opto/gcm.cpp

```diff
@@ -769,6 +769,30 @@ Block* PhaseCFG::insert_anti_dependences(Block* LCA, Node* load, bool verify) {
     }
   }
   worklist_def_use_mem_states.push(nullptr, initial_mem);
+  Block* initial_mem_block = get_block_for_node(initial_mem);
+  if (load->in(0) && initial_mem_block != nullptr) {
+    // If the load has an explicit control input and initial_mem has a block,
+    // walk up the dominator tree from the early block to the initial memory
+    // block. If we in a block find memory Phi(s) that can alias initial_mem,
+    // these are also potential initial memory states and there may be further
+    // required anti dependences due to them.
+    assert(initial_mem_block->dominates(early), "invariant");
+    Block* b = early;
+    // Stop searching when we run out of dominators (b == nullptr) or when we
+    // step past the initial memory block (b == initial_mem_block->_idom).
+    while (b != nullptr && b != initial_mem_block->_idom) {
+      if (b == initial_mem_block && !initial_mem->is_Phi()) {
+        break;
+      }
+      for (uint i = 0; i < b->number_of_nodes(); ++i) {
+        Node* n = b->get_node(i);
+        if (n->is_memory_phi() && C->can_alias(n->adr_type(), load_alias_idx)) {
+          worklist_def_use_mem_states.push(nullptr, n);
```

Comment:
We may push `<nullptr, initial_mem>` again here, is that an issue? If not, maybe add a comment explaining why.